soroban-env-host/src/budget/dimension.rs

```diff
@@ -0,0 +1,176 @@
+    /// of comparing to limit.
+    pub(crate) total_count: u64,
+
+    /// The internal limit tracks work done internally that is not exposed to the
+    /// external user -- it does not affect fees or decide the invocation outcome
+    /// in any way (no error due to exceeding the internal limit). It exists solely
+    /// for dos prevention. Such work include diagnostic logging, or work that
+    /// exists only for preflight.
+    pub(crate) internal_limit: u64,
+
+    /// Similar to `total_count`, but towards the `internal_limit`
+    pub(crate) internal_total_count: u64,
+}
+
+impl Debug for BudgetDimension {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        writeln!(
+            f,
+            "limit: {}, total_count: {}",
+            self.limit, self.total_count
+        )?;
+
+        for ct in ContractCostType::variants() {
+            writeln!(f, "CostType {:?}, count {}", ct, self.counts[ct as usize])?;
+            writeln!(f, "model: {:?}", self.cost_models[ct as usize])?;
+        }
+
+        writeln!(
+            f,
+            "internal limit: {}, internal_total_count: {}",
+            self.internal_limit, self.internal_total_count
+        )?;
+        Ok(())
+    }
+}
+
+impl BudgetDimension {
+    pub(crate) fn new() -> Self {
+        let mut bd = BudgetDimension::default();
+        for _ct in ContractCostType::variants() {
+            bd.cost_models.push(MeteredCostComponent {
+                const_term: 0,
+                lin_term: ScaledU64(0),
+            });
+            bd.counts.push(0);
+        }
+        bd
+    }
+
+    pub(crate) fn try_from_config(cost_params: ContractCostParams) -> Result<Self, HostError> {
+        let cost_models = cost_params
+            .0
+            .iter()
+            .map(|p| MeteredCostComponent::try_from(p))
+            .collect::<Result<Vec<MeteredCostComponent>, HostError>>()?;
+
+        Ok(Self {
+            cost_models,
+            limit: Default::default(),
+            counts: vec![0; cost_params.0.len()],
+            total_count: Default::default(),
+            internal_limit: Default::default(),
+            internal_total_count: Default::default(),
+        })
+    }
+
+    pub(crate) fn get_cost_model(&self, ty: ContractCostType) -> &MeteredCostComponent {
+        &self.cost_models[ty as usize]
+    }
+
+    pub(crate) fn get_cost_model_mut(&mut self, ty: ContractCostType) -> &mut MeteredCostComponent {
+        &mut self.cost_models[ty as usize]
+    }
+
+    pub(crate) fn get_total_count(&self) -> u64 {
+        self.total_count
+    }
+
+    pub(crate) fn get_remaining(&self) -> u64 {
+        self.limit.saturating_sub(self.total_count)
+    }
+
+    pub(crate) fn reset(&mut self, limit: u64) {
+        self.limit = limit;
+        self.total_count = 0;
+        for v in &mut self.counts {
+            *v = 0;
+        }
+        self.internal_limit = limit;
```

Comment:
I think we might want to expose the internal limit multiplier or maybe just allow setting it to a flat value.

Reply:
Yeah we could have a host constant for the limit multiplier or a flat value. I don't feel too strongly about that. 
@graydon suggested keeping things minimal and when needed it should be easy to switch to an internal constant.

Reply:
yeah I don't see any need to overthink this: it's _specifically_ an unobservable part of the semantics, so if it has a simple interface now and someone asks us to make it a more-complex interface sometime in the future, it should not effect anything.

Reply:
I think there is an immediate requirement for preflight (soroban-rpc) to be able to set this up.

Reply:
@dmkozh I think adding a couple of setter functions in the host should suffice? Or do we need platform team's input on this.

Reply:
Just a public setter is sufficient for now.

Reply:
Added pub setters for the shadow budget limits.